// @storybook/ui/dist/components/sidebar/Search.stories.js
"use strict";

require("core-js/modules/es.object.assign");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.filledIn = exports.focussed = exports.simple = exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonActions = require("@storybook/addon-actions");

var _Search = require("./Search");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var _default = {
  component: _Search.Search,
  title: 'UI/Sidebar/Search',
  decorators: [function (storyFn) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      style: {
        width: '240px'
      }
    }, storyFn());
  }]
};
exports["default"] = _default;
var actions = (0, _addonActions.actions)('onChange');
var pureActions = Object.assign(Object.assign({}, actions), (0, _addonActions.actions)('onSetFocussed'));

var simple = function simple() {
  return /*#__PURE__*/_react["default"].createElement(_Search.Search, actions);
};

exports.simple = simple;
simple.displayName = "simple";

var focussed = function focussed() {
  return /*#__PURE__*/_react["default"].createElement(_Search.Search, _extends({
    defaultFocussed: true
  }, pureActions));
};

exports.focussed = focussed;
focussed.displayName = "focussed";

var filledIn = function filledIn() {
  return /*#__PURE__*/_react["default"].createElement(_Search.Search, _extends({
    defaultValue: "Searchstring"
  }, pureActions));
};

exports.filledIn = filledIn;
filledIn.displayName = "filledIn";
