// @storybook/ui/dist/components/sidebar/Tree/components.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.array.index-of");

require("core-js/modules/es.function.name");

require("core-js/modules/es.object.assign");

require("core-js/modules/es.object.keys");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultLink = exports.DefaultRootTitle = exports.DefaultHead = exports.DefaultLeaf = exports.LeafStyle = exports.DefaultMessage = exports.prevent = exports.DefaultFilter = exports.A = exports.DefaultList = exports.DefaultSection = void 0;

var _react = _interopRequireDefault(require("react"));

var _theming = require("@storybook/theming");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var DefaultSection = _theming.styled.div();

exports.DefaultSection = DefaultSection;

var DefaultList = _theming.styled.div();

exports.DefaultList = DefaultList;

var A = _theming.styled.a();

exports.A = A;
var DefaultFilter = (0, _theming.styled)(function (props) {
  return /*#__PURE__*/_react["default"].createElement("input", _extends({
    placeholder: "search..."
  }, props));
})({
  width: '100%',
  background: 'transparent',
  border: '1px solid black'
});
exports.DefaultFilter = DefaultFilter;

var prevent = function prevent(e) {
  e.preventDefault();
  return false;
};

exports.prevent = prevent;

var DefaultMessage = _theming.styled.div({});

exports.DefaultMessage = DefaultMessage;

var LeafStyle = _theming.styled.div({
  minHeight: 24,
  display: 'flex',
  alignItems: 'center',
  flex: 1
}, function (_ref) {
  var depth = _ref.depth;
  return {
    paddingLeft: depth * 10
  };
}, function (_ref2) {
  var isSelected = _ref2.isSelected;
  return {
    background: isSelected ? '#CFD8DC' : 'transparent'
  };
});

exports.LeafStyle = LeafStyle;

var DefaultLeaf = function DefaultLeaf(_ref3) {
  var name = _ref3.name,
      isSelected = _ref3.isSelected,
      depth = _ref3.depth,
      rest = _objectWithoutProperties(_ref3, ["name", "isSelected", "depth"]);

  return /*#__PURE__*/_react["default"].createElement(LeafStyle, _extends({
    isSelected: isSelected,
    depth: depth
  }, rest), name);
};

exports.DefaultLeaf = DefaultLeaf;
DefaultLeaf.displayName = "DefaultLeaf";

var DefaultHead = function DefaultHead(_ref4) {
  var name = _ref4.name,
      depth = _ref4.depth,
      _ref4$isExpanded = _ref4.isExpanded,
      isExpanded = _ref4$isExpanded === void 0 ? false : _ref4$isExpanded,
      _ref4$isSelected = _ref4.isSelected,
      isSelected = _ref4$isSelected === void 0 ? false : _ref4$isSelected,
      _ref4$isComponent = _ref4.isComponent,
      isComponent = _ref4$isComponent === void 0 ? false : _ref4$isComponent;
  return /*#__PURE__*/_react["default"].createElement(LeafStyle, {
    isSelected: isSelected,
    depth: depth
  }, /*#__PURE__*/_react["default"].createElement("span", null, isExpanded ? '-' : '+', isComponent ? '!' : ''), /*#__PURE__*/_react["default"].createElement("span", null, name));
};

exports.DefaultHead = DefaultHead;
DefaultHead.displayName = "DefaultHead";

var DefaultRootTitle = _theming.styled.h4({});

exports.DefaultRootTitle = DefaultRootTitle;

var DefaultLink = function DefaultLink(_ref5) {
  var id = _ref5.id,
      prefix = _ref5.prefix,
      children = _ref5.children,
      rest = _objectWithoutProperties(_ref5, ["id", "prefix", "children"]);

  return /*#__PURE__*/_react["default"].createElement(A, _extends({
    href: "#!".concat(prefix).concat(id)
  }, rest, {
    onClick: function onClick(e) {
      return prevent(e) || rest.onClick(e);
    }
  }), children);
};

exports.DefaultLink = DefaultLink;
DefaultLink.displayName = "DefaultLink";
