// @storybook/ui/dist/components/sidebar/Tree/ListItem.stories.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

require("core-js/modules/es.object.assign");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NestedDepths = exports.WithLongName = exports.StorySelected = exports.Story = exports.ComponentExpanded = exports.Component = exports.Group = exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _ListItem = require("./ListItem");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var _default = {
  component: _ListItem.ListItem,
  title: 'UI/Sidebar/ListItem',
  decorators: [function (storyFn) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      style: {
        width: '240px'
      }
    }, storyFn());
  }]
};
exports["default"] = _default;
var baseProps = {
  depth: 0,
  id: 'foo-bar',
  isComponent: false,
  isLeaf: false,
  isExpanded: false,
  isSelected: false,
  kind: 'foo',
  name: 'bar',
  refId: ''
};

var Group = function Group() {
  return /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, baseProps);
};

exports.Group = Group;
Group.displayName = "Group";

var Component = function Component() {
  return /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, _extends({}, baseProps, {
    isComponent: true
  }));
};

exports.Component = Component;
Component.displayName = "Component";

var ComponentExpanded = function ComponentExpanded() {
  return /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, _extends({}, baseProps, {
    isComponent: true,
    isExpanded: true
  }));
};

exports.ComponentExpanded = ComponentExpanded;
ComponentExpanded.displayName = "ComponentExpanded";

var Story = function Story() {
  return /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, _extends({}, baseProps, {
    isLeaf: true
  }));
};

exports.Story = Story;
Story.displayName = "Story";

var StorySelected = function StorySelected() {
  return /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, _extends({}, baseProps, {
    isLeaf: true,
    isSelected: true
  }));
};

exports.StorySelected = StorySelected;
StorySelected.displayName = "StorySelected";

var WithLongName = function WithLongName() {
  return /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, _extends({}, baseProps, {
    name: "Holy cow this is a very very very long name to be in this tiny area."
  }));
};

exports.WithLongName = WithLongName;
WithLongName.displayName = "WithLongName";

var NestedDepths = function NestedDepths() {
  return /*#__PURE__*/_react["default"].createElement(_react.Fragment, null, /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, _extends({}, baseProps, {
    depth: 1
  })), /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, _extends({}, baseProps, {
    depth: 2
  })), /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, _extends({}, baseProps, {
    depth: 3
  })), /*#__PURE__*/_react["default"].createElement(_ListItem.ListItem, _extends({}, baseProps, {
    depth: 4
  })));
};

exports.NestedDepths = NestedDepths;
NestedDepths.displayName = "NestedDepths";
