// @storybook/addon-docs/dist/frameworks/react/__testfixtures__/8140-js-prop-types-oneof/input.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.component = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Alert = function Alert(props) {
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, JSON.stringify(props));
};

Alert.defaultProps = {
  mode: 'static',
  type: 'warning'
};
Alert.propTypes = {
  mode: _propTypes["default"].oneOf(['static', 'timed']),
  type: _propTypes["default"].oneOf(['success', 'warning', 'error', 'primary']),
  message: _propTypes["default"].string.isRequired,

  /**
   * No background or border if static alert
   */
  blank: _propTypes["default"].bool,

  /**
   * Allows icon override, accepts material icon name
   */
  icon: _propTypes["default"].string
};
var component = Alert;
exports.component = component;
