// @storybook/components/dist/Loader/Loader.js
"use strict";

require("core-js/modules/es.object.assign");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Loader = void 0;

var _react = _interopRequireDefault(require("react"));

var _theming = require("@storybook/theming");

var _animation = require("../shared/animation");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var LoaderWrapper = _theming.styled.div(function (_ref) {
  var _ref$size = _ref.size,
      size = _ref$size === void 0 ? 32 : _ref$size;
  return {
    borderRadius: '50%',
    cursor: 'progress',
    display: 'inline-block',
    overflow: 'hidden',
    position: 'absolute',
    transition: 'all 200ms ease-out',
    verticalAlign: 'top',
    top: '50%',
    left: '50%',
    marginTop: -(size / 2),
    marginLeft: -(size / 2),
    height: size,
    width: size,
    zIndex: 4,
    borderWidth: 2,
    borderStyle: 'solid',
    borderColor: 'rgba(97, 97, 97, 0.29)',
    borderTopColor: 'rgb(100,100,100)',
    animation: "".concat(_animation.rotate360, " 0.7s linear infinite"),
    mixBlendMode: 'difference'
  };
});

var Loader = function Loader(props) {
  return /*#__PURE__*/_react["default"].createElement(LoaderWrapper, _extends({
    "aria-label": "Content is loading ...",
    "aria-live": "polite",
    role: "status"
  }, props));
};

exports.Loader = Loader;
Loader.displayName = "Loader";
